### PR TITLE
Added second layer of filtering

### PR DESCRIPTION
**Pull request by Alamgir-K**
## Context

 The current implementation recommended five cars without making use of senso rate. Now, we get a list of 10 cars and then further filter them by choosing only 5 of them with the lowest interest rates as returned by making a call to the senso api
 #10 

## Your Changes

**Description**:
 The CarRecommendations package now has 4 files. We first use EstimateBudget to get a user's budget, then use that budget to filter for 10 possible cars through BudgetFilter, then pass these 10 cars into InterestFilter using SensoRate on each car to find the best five cars with the lowest interest rates

**Type of change** (select all that apply):

- [x] New feature (additional functionality separate from existing code)
- [x] Refactoring (changing code structure only)
- [x] Breaking change (changing code functionality)
- [x] Bug fix (non-breaking change that solves an error/issue)
- [x] User Interface change (design changes)

## Testing

 Code was tested by calling our /cars endpoint through Postman and checking if the correct Json is being returned.

## Notes

 Everything works properly but now we will create two branches off of this to add documentation and unit testing

## Checklist

- [x] Code is up to date with main branch (no merge conflicts)
- [ ] One or two people have reviewed this pull request

### CODE REVIEW

File: data/Car_Data.csv [hunk 1]
```diff
@@ -1,2110 +1,110 @@
 vin,price,brand,model,year,image
```
Comment (leenlab2) [inline data/Car_Data.csv:1]: Why was this deleted?

File: src/main/java/com/aviva/CarRecommendations/SensoRate.java [hunk 2]
```diff
@@ -10,44 +11,41 @@
...
-    public static void main(String[] args) throws IOException, InterruptedException {
-        AccountHolder user;
-        AccountHolderDataInterface accountData = new CSVAccountHolderData();
-        user = accountData.getClientByID("5002357538983918");
-        System.out.println(getApprovedLoanAmount(user));
+        try {
+            // use the client to send the request
+            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+            JSONObject responseJSON = new JSONObject(response.body());
+            return responseJSON.getFloat("interestRate");
```
Comment (leenlab2) [inline src/main/java/com/aviva/CarRecommendations/SensoRate.java:43]: We need more than just the interest rate from Senso

File: src/main/java/com/aviva/DataAccess/BankingDataProcess.java [hunk 3]
```diff
@@ -57,9 +59,16 @@ public int getCreditScore(String accountNumber){
         try {
             latestCreditResultSet.next();
             creditScore = latestCreditResultSet.getInt("creditScore");
+            return creditScore;
         } catch (SQLException e) {
             System.out.println("Could not get credit score");
         }
         return creditScore;
```
Comment (leenlab2) [inline src/main/java/com/aviva/DataAccess/BankingDataProcess.java:66]: remove this, keep the one inside try

File: src/main/java/com/aviva/DataAccess/BankingDataProcess.java [hunk 4]
```diff
@@ -57,9 +59,16 @@ public int getCreditScore(String accountNumber){
...
     }
+
+    public AccountHolder getAccountHolder(String accountNumber) {
```
Comment (leenlab2) [inline src/main/java/com/aviva/DataAccess/BankingDataProcess.java:69]: this method should probably be renamed

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import com.aviva.Entities.Car;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class InterestFilter {
+
+    public JSONObject getBestFiveCars(String accountNumber) {
```
Comment (leenlab2) [inline src/main/java/com/aviva/CarRecommendations/InterestFilter.java:14]: this method is pretty long, we should probably split it up a bit. I've included some suggestions

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        BankingDataProcess bdpInit = new BankingDataProcess();
+        AccountHolder user = bdpInit.getAccountHolder(accountNumber);
+
+        EstimateBudget ebInit = new EstimateBudget();
+        float yearlyBudget = ebInit.calculateYearlyBudget(user);
+        user.setSavings(yearlyBudget);
```
Comment (leenlab2) [inline src/main/java/com/aviva/CarRecommendations/InterestFilter.java:20]: fyi, savings is going to get removed when An merges her entity change, so maybe add a TODO to remind us to fix this later

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class InterestFilter {
+
+    public JSONObject getBestFiveCars(String accountNumber) {
+        BankingDataProcess bdpInit = new BankingDataProcess();
+        AccountHolder user = bdpInit.getAccountHolder(accountNumber);
```
Comment (leenlab2) [inline src/main/java/com/aviva/CarRecommendations/InterestFilter.java:16]: maybe move this to the constructor of AccountHolder

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+        EstimateBudget ebInit = new EstimateBudget();
+        float yearlyBudget = ebInit.calculateYearlyBudget(user);
+        user.setSavings(yearlyBudget);
+        float monthlyBudget = ebInit.calculateMonthlyBudget(user);
+        user.setMonthlyBudget(monthlyBudget);
```
Comment (leenlab2) [inline src/main/java/com/aviva/CarRecommendations/InterestFilter.java:22]: the budgeting stuff can also be moved to the constructor of accountholder

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            float interestRate = srInit.getInterestRate(user, cars.get(i));
+            rates[i] = interestRate;
+        }
+
+        float[] sortedRates = Arrays.copyOf(rates, 10);
+        Arrays.sort(sortedRates);
+
+        for (int i = 0; i < 5; i++) {
+            float sortedRate = sortedRates[i];
+            int carIndex = getFirstIndex(rates, sortedRate);
```
Comment (leenlab2) [inline src/main/java/com/aviva/CarRecommendations/InterestFilter.java:41]: I feel like this could have been done in a slightly better way. Have you considered using a HashMap?

File: src/main/java/com/aviva/CarRecommendations/SensoRate.java (hunk 2 above)
Comment (leenlab2) [reply src/main/java/com/aviva/CarRecommendations/SensoRate.java:43]: if you can save the installment information, or at least how much we are paying for that car per month

**Review by leenlab2 — COMMENTED** [review]
I left some specific comments throughout the code, but as a general comment maybe improve the names of the budget objects? bfInit is a bit confusing, for instance. name them something like budgetFilter

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java (hunk 9 above)
Comment (nguya03584) [reply src/main/java/com/aviva/CarRecommendations/InterestFilter.java:41]: I agree with Leen's on this because HashMap might be more efficient.

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            bestFive.add(cars.get(carIndex));
+            rates[carIndex] = (float) -1.0;
+        }
+
+        // Converting to JSON
```
Comment (nguya03584) [inline src/main/java/com/aviva/CarRecommendations/InterestFilter.java:46]: I think we should put a code converting the Car objects to JSONObject into its own method, so it will this method less long.

**Review by nguya03584 — COMMENTED** [review]
Besides agreeing with what Leen already suggested, I added another suggestion to shorten the getBestFiveCar method.

File: src/main/java/com/aviva/CarRecommendations/SensoRate.java (hunk 2 above)
Comment (Alamgir-K) [reply src/main/java/com/aviva/CarRecommendations/SensoRate.java:43]: we have made method to get a car's installments so you can use this however need be

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java (hunk 5 above)
Comment (Alamgir-K) [reply src/main/java/com/aviva/CarRecommendations/InterestFilter.java:14]: we can't realistically shorten it aside from few efficiency changes which we dont have time for right now

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java (hunk 8 above)
Comment (Alamgir-K) [reply src/main/java/com/aviva/CarRecommendations/InterestFilter.java:22]: doesn't make sense to do this atm because sometimes we will initialize an account holder without a credit score and at a later point get their latest so for now we should leave it be

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java (hunk 9 above)
Comment (Alamgir-K) [reply src/main/java/com/aviva/CarRecommendations/InterestFilter.java:41]: efficiency can be dealt with later when we have more time

File: src/main/java/com/aviva/CarRecommendations/InterestFilter.java (hunk 10 above)
Comment (Alamgir-K) [reply src/main/java/com/aviva/CarRecommendations/InterestFilter.java:46]: we can leave it be for now and make it look cleaner for phase 2
